File: test/rustty/host_queries.py
"""Explicit host-query decisions and their terminal protocol replies."""


def requests():
    def case(name, operations, host=None):
        return ({"id": "protocol/host/" + name,
                 "operations": [op if isinstance(op, dict) else {"op": "write", "data": op.hex()} for op in operations],
                 "host": host or {}}, ["effects.host", "effects.pty"])

    queries = b"\x1b[c\x1b[>c\x1b[=c\x1b[>q\x05\x1b[?996n\x1b[14t\x1b[16t\x1b[18t"
    populated = {"device_attributes": {}, "color_scheme": "dark", "size": {},
                 "xtversion": b"rustty test".hex(), "enquiry": b"answer".hex()}
    yield case("defaults/absent", [queries])
    yield case("defaults/configured", [queries], populated)
    yield case("defaults/changed", [queries, {"op": "host_options", "host": populated}, queries,
                                     {"op": "host_options", "host": {}}, queries])
    yield case("defaults/ris", [queries, b"\x1bc", queries], populated)
    yield case("defaults/direct-reset", [queries, {"op": "terminal_reset"}, queries], populated)

    for index, attributes in enumerate(({}, {"features": []}, {"conformance_level": 1, "features": [1, 6, 22]},
                                         {"conformance_level": 65535, "features": [0, 65535, 22, 22], "device_type": 65535,
                                          "firmware_version": 65535, "rom_cartridge": 65535, "unit_id": 4294967295},
                                         {"features": [65535] * 256, "unit_id": 0xAABBCCDD})):
        yield case(f"attributes/values/{index}", [b"\x1b[c\x1b[>0c\x1b[=0c"], {"device_attributes": attributes})
    for index, query in enumerate((b"c", b"0c", b"1c", b"0;0c", b">c", b">0c", b">1c", b">0;0c",
                                    b"=c", b"=0c", b"=1c", b"?c", b"?0c", b"0:c", b">65535c")):
        yield case(f"attributes/context/{index}", [b"\x1b[" + query], {"device_attributes": {}})
    yield case("attributes/decid", [b"\x1bZ"], {"device_attributes": {}})
    yield case("attributes/decid-absent", [b"\x1bZ"])

    for name, key, command, lengths in (("version", "xtversion", b"\x1b[>0q", (0, 1, 255, 256, 257, 280, 281, 282, 512)),
                                        ("enquiry", "enquiry", b"\x05", (0, 1, 254, 255, 256, 257))):
        for length in lengths:
            yield case(f"{name}/length/{length}", [command], {key: (b"x" * length).hex()})
        yield case(f"{name}/raw-bytes", [command], {key: b"a\x00\xff\x1b\\".hex()})
    for index, query in enumerate((b">q", b">0q", b">1q", b">0;0q", b"q", b"?q", b"=q")):
        yield case(f"version/context/{index}", [b"\x1b[" + query], {"xtversion": b"version".hex()})
    for index, data in enumerate((b"\x05", b"\x1b[\x05m", b"\x1b]\x05\x07", b"\x1bP\x05q\x1b\\")):
        yield case(f"enquiry/context/{index}", [data], {"enquiry": b"answer".hex()})

    for scheme in (None, "none", "dark", "light"):
        yield case(f"color-scheme/{scheme}", [b"\x1b[?996n\x1b[996n\x1b[?997n\x1b[?996;0n"], {"color_scheme": scheme})
    for private in (b"", b"?"):
        for value in (5, 6, 996, 998, 0, 65535):
            yield case(f"status/{private.hex()}/{value}", [b"\x1b[" + private + str(value).encode() + b"n"], populated)

    sizes = [None, {}, {"available": False}, {"rows": 0, "columns": 0, "cell_width": 0, "cell_height": 0},
             {"rows": 65535, "columns": 65535, "cell_width": 4294967295, "cell_height": 4294967295}]
    for index, size in enumerate(sizes):
        yield case(f"size/values/{index}", [b"\x1b[14t\x1b[16t\x1b[18t\x1b[?2048h\x1b[?2048h\x1b[?2048l"], {"size": size})
    for index, query in enumerate((b"14t", b"14;0t", b"14;1t", b"14;2t", b"16t", b"18t", b"18;1t", b"21t", b"?14t")):
        yield case(f"size/context/{index}", [b"\x1b[" + query], {"size": {}})
    yield case("size/saved-mode", [b"\x1b[?2048h\x1b[?2048s\x1b[?2048l\x1b[?2048r\x1b[?2048$p"], {"size": {}})
    for enabled in (False, True):
        for cell_size in (None, [0, 0], [8, 16], [4294967295, 4294967295]):
            yield case(f"size/resize/{enabled}/{cell_size}", [
                b"\x1b[?2048h" if enabled else b"\x1b[?2048l",
                {"op": "resize", "cols": 12, "rows": 4, "cell_size": cell_size},
                {"op": "resize", "cols": 10, "rows": 5, "cell_size": cell_size},
                b"\x1b[?2048$p",
            ])

    for visible in (False, True):
        yield case(f"visibility/{visible}", [b"\x1b[?998n\x1b[?2033$p\x1b[?2033h\x1b[?2033h\x1b[?2033l",
                                              b"\x1bc\x1b[?998n", {"op": "terminal_reset"}, b"\x1b[?998n"], {"visible": visible})
    yield case("visibility/change", [b"\x1b[?2033h", {"op": "host_options", "host": {"visible": False}}, b"\x1b[?998n"])
    yield case("visibility/saved-mode", [b"\x1b[?2033h\x1b[?2033s\x1b[?2033l\x1b[?2033r\x1b[?2033$p"], {"visible": False})
    for enabled in (False, True):
        yield case(f"title-report/{enabled}", [b"\x1b]2;Title\x07\x1b[21t\x1bc\x1b[21t"], {"title_report": enabled})

    for index, name in enumerate((None, b"", b"rustty", b"xterm-rustty", "héllo".encode(), b"raw\x00\xff",
                                  b"x" * 128, b"x" * 129, b"name with punctuation;=")):
        query = b"\x1bP+q544e\x1b\\"
        yield case(f"terminfo/{index}", [query, b"\x1bc", query, {"op": "terminal_reset"}, query],
                   {"terminfo_name": name.hex() if name is not None else None})

File: crates/rustty-vt/tests/host_queries.rs
use rustty_vt::query::{ColorScheme, DeviceAttributes, Size, SizeStyle};
use rustty_vt::{Effect, EffectHandler, Terminal};

#[derive(Default)]
struct Readonly(Vec<Effect>);
impl EffectHandler for Readonly {
    fn effect(&mut self, effect: Effect) {
        self.0.push(effect);
    }
}

#[test]
fn linefeed_mode_effects_keep_stream_order_including_reset() {
    let stream = b"\x1b[20h\x1b[5n\x1b[20l\x1b[5n\x1bc";
    for split in 0..=stream.len() {
        let mut terminal = Terminal::new(10, 2, 0);
        terminal.linefeed_mode_events = true;
        let mut effects = terminal.feed(&stream[..split]);
        effects.extend(terminal.feed(&stream[split..]));
        assert_eq!(
            effects,
            [
                Effect::LinefeedMode(true),
                Effect::Write(b"\x1b[0n".to_vec()),
                Effect::LinefeedMode(false),
                Effect::Write(b"\x1b[0n".to_vec()),
                Effect::LinefeedMode(false),
                Effect::Progress {
                    state: 0,
                    value: None
                },
            ]
        );
        assert!(terminal.linefeed_mode_events);
    }
}

#[test]
fn focus_reporting_sends_current_host_state_on_enable_and_restore() {
    for focused in [false, true] {
        let mut terminal = Terminal::new(10, 2, 0);
        assert!(terminal.feed(b"\x1b[?1004h").is_empty());
        terminal.query_defaults.focused = Some(focused);
        let report = Effect::Write(if focused { b"\x1b[I" } else { b"\x1b[O" }.to_vec());
        for sequence in [
            b"\x1b[?1004h".as_slice(),
            b"\x1b[?1004s\x1b[?1004l\x1b[?1004r",
        ] {
            assert_eq!(
                terminal.feed(sequence).as_slice(),
                std::slice::from_ref(&report)
            );
        }
        terminal.feed(b"\x1bc");
        assert_eq!(terminal.feed(b"\x1b[?1004h"), [report]);
    }
}

#[test]
fn synchronous_hosts_are_readonly_unless_callbacks_supply_answers() {
    let mut terminal = Terminal::new(10, 2, 0);
    let mut host = Readonly::default();
    terminal.feed_with_handler(
        b"\x05\x1bZ\x1b[c\x1b[>c\x1b[=c\x1b[?996n\x1b[14t\x1b[16t\x1b[18t\x1b[>q",
        &mut host,
    );
    assert_eq!(host.0, [Effect::Write(b"\x1bP>|libghostty\x1b\\".to_vec())]);
}

#[test]
fn host_queries_use_binary_answers_and_preserve_callback_order() {
    #[derive(Default)]
    struct Host {
        calls: Vec<&'static str>,
        writes: Vec<Vec<u8>>,
    }
    impl EffectHandler for Host {
        fn effect(&mut self, effect: Effect) {
            if let Effect::Write(bytes) = effect {
                self.calls.push("write");
                self.writes.push(bytes);
            }
        }
        fn enquiry(&mut self) -> Vec<u8> {
            self.calls.push("enquiry");
            vec![0, 255]
        }
        fn xtversion(&mut self) -> Vec<u8> {
            self.calls.push("version");
            b"custom\xff".to_vec()
        }
        fn color_scheme(&mut self) -> Option<ColorScheme> {
            self.calls.push("scheme");
            Some(ColorScheme::Light)
        }
        fn device_attributes(&mut self) -> Option<DeviceAttributes> {
            self.calls.push("attributes");
            Some(DeviceAttributes {
                conformance_level: 65,
                features: vec![22, 52],
                device_type: 64,
                firmware_version: 123,
                rom_cartridge: 2,
                unit_id: 0xabcdef01,
            })
        }
        fn size(&mut self) -> Option<Size> {
            self.calls.push("size");
            Some(Size {
                rows: 12,
                columns: 34,
                cell_width: 9,
                cell_height: 18,
            })
        }
    }
    let mut terminal = Terminal::new(10, 2, 0);
    let mut host = Host::default();
    terminal.feed_with_handler(
        b"\x05\x1b[>q\x1b[?996n\x1b[c\x1b[>c\x1b[=c\x1b[14t\x1b[?2048h",
        &mut host,
    );
    assert_eq!(
        host.calls,
        [
            "enquiry",
            "write",
            "version",
            "write",
            "scheme",
            "write",
            "attributes",
            "write",
            "attributes",
            "write",
            "attributes",
            "write",
            "size",
            "write",
            "size",
            "write"
        ]
    );
    assert_eq!(
        host.writes,
        [
            vec![0, 255],
            b"\x1bP>|custom\xff\x1b\\".to_vec(),
            b"\x1b[?997;2n".to_vec(),
            b"\x1b[?65;22;52c".to_vec(),
            b"\x1b[>64;123;2c".to_vec(),
            b"\x1bP!|ABCDEF01\x1b\\".to_vec(),
            b"\x1b[4;216;306t".to_vec(),
            b"\x1b[48;12;34;216;306t".to_vec(),
        ]
    );
}

#[test]
fn host_defaults_survive_ris_and_are_reapplied_after_snapshot_restore() {
    let mut terminal = Terminal::new(10, 2, 0);
    assert_eq!(
        terminal.feed(b"\x1b[>0q"),
        [Effect::Write(
            concat!("\x1bP>|ghostty ", env!("CARGO_PKG_VERSION"), "\x1b\\")
                .as_bytes()
                .to_vec()
        )]
    );
    terminal.query_defaults.xtversion = b"embedder".to_vec();
    terminal.query_defaults.enquiry = b"answer".to_vec();
    terminal.query_defaults.color_scheme = Some(ColorScheme::Dark);
    terminal.query_defaults.device_attributes = None;
    terminal.title_report = true;
    terminal.visible = false;
    terminal.terminfo_name = Some(vec![b'r', 255]);
    terminal.feed(b"\x1bc\x1b]2;title\x07");
    assert_eq!(
        terminal.feed(b"\x05\x1b[c\x1b[>q\x1b[?996n\x1b[?998n\x1b[21t"),
        [
            Effect::Write(b"answer".to_vec()),
            Effect::Write(b"\x1bP>|embedder\x1b\\".to_vec()),
            Effect::Write(b"\x1b[?997;1n".to_vec()),
            Effect::Write(b"\x1b[?999;2n".to_vec()),
            Effect::Write(b"\x1b]ltitle\x1b\\".to_vec()),
        ]
    );
    let bytes = rustty_vt::snapshot::encode_to_vec(&terminal).unwrap();
    let restored = rustty_vt::snapshot::decode(bytes.as_slice(), Default::default()).unwrap();
    assert_ne!(restored.query_defaults, terminal.query_defaults);
    assert!(!restored.title_report);
    assert!(restored.visible);
    assert_eq!(restored.terminfo_name, None);
    assert_eq!(terminal.terminfo_name, Some(vec![b'r', 255]));
}

#[test]
fn resize_reports_complete_geometry_without_truncating_wire_values() {
    let mut terminal = Terminal::new(10, 2, 0);
    terminal.modes.set(true, 2048, true);
    assert!(terminal.resize_with_cell_size(10, 2, None).is_empty());
    assert_eq!(
        terminal.resize_with_cell_size(10, 2, Some((9, 18))),
        [Effect::Write(b"\x1b[48;2;10;36;90t".to_vec())]
    );
    terminal.modes.set(true, 2026, true);
    let report = terminal.resize_with_cell_size(10, 2, Some((u32::MAX, u32::MAX)));
    assert_eq!(
        report,
        [Effect::Write(
            b"\x1b[48;2;10;8589934590;42949672950t".to_vec()
        )]
    );
    assert_eq!(
        (terminal.width_px, terminal.height_px),
        (u32::MAX, u32::MAX)
    );
    assert!(!terminal.modes.dec(2026));
    assert_eq!(
        Size {
            rows: u16::MAX,
            columns: u16::MAX,
            cell_width: u32::MAX,
            cell_height: u32::MAX
        }
        .encode(SizeStyle::TextPixels),
        b"\x1b[4;281470681677825;281470681677825t"
    );
}
